### ISSUE TITLE
Functions with same intersection and conditional type in parameter list not assignable to each other

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#32442]
**TypeScript Version:**  3.5.1

**Search Terms:**

**Code**

```ts
/**
 * We should never be able to create a value of this type legitimately.
 * 
 * `ErrorMessageT` is our error message
 */
interface CompileError<ErrorMessageT extends any[]> {
  /**
   * There should never be a value of this type
   */
  readonly __compileError : never;
}
type ErrorA = CompileError<["I am error A"]>;
type ErrorB = CompileError<["I am error B"]>;

declare const errorA : ErrorA;
/**
 * Different compile errors are assignable to each other.
 */
const errorB : ErrorB = errorA;

/**
 * Pretend this is `v1.0.0` of your library.
 */
declare function foo <N extends number> (
  /**
   * This is how we use `CompileError<>` to prevent `3` from being
   * a parameter
   */
  n : (
    N &
    (Extract<3, N> extends never ?
    unknown :
    CompileError<[3, "is not allowed; received", N]>)
  )
) : void;

/**
 * Argument of type '3' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3]>'.
 */
foo(3);
/**
 * OK!
 */
foo(5);
/**
 * Argument of type '3 | 5' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3 | 5]>'.
 */
foo(5 as 3|5);
/**
 * Argument of type 'number' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", number]>'.
 */
foo(5 as number);

///////////////////////////////////////////////////////////////////

/**
 * The same as `foo<>()` but with a different error message.
 * 
 * Pretend this is `v1.1.0` of your library.
 */
declare function bar <N extends number> (
  n : (
    N &
    (Extract<3, N> extends never ?
    unknown :
    CompileError<[3, "is not allowed; received", N]>)
  )
) : void;

/**
 * Expected: Assignable to each other
 * Actual: Not assignable to each other
 */
const fooIsAssignableToBar : typeof bar = foo;
const barIsAssignableToFoo : typeof foo = bar;
```

**Expected behavior:**

The following should have no errors,
```ts
const fooIsAssignableToBar : typeof bar = foo;
const barIsAssignableToFoo : typeof foo = bar;
```

**Actual behavior:**

It has errors

**Playground Link:** 

[Playground](http://www.typescriptlang.org/play/#code/PQKhFgCgAIWh1AptAzgCwPYFcA2ATaAO0QDdEAnaAI2QEMqdkAXDaAY3MVqbuhNpxZkGAGbQmaAJYpxATwAOyRgHNJTSQFtuiHLIB0UWNENwABgFFy5DOQCyiFClrLEAFVPRp0bJQrXKGg5OLibAUJKEPOQitGzIAMIYGvKSjJb+ADzpNvaOzm7QiAAePIR4MrSEsgDaALoAfNAA3obQoBAw0EauaBTI6Nj4RKQU1Lz8gsJiEl5MCoitsGGdnLR4GIS60AD622xJKWlWNtAAXMNk5ADcUAC+UHOK0NnkAILQALzQicmpiC8ZaoAIgAktBaBpCsdKK8gQ0bpBHsgXgAhT7fA5-AHAsEQqH+aAouH1BFQPCINg4WicdgbFBMfE2d7nF6vBHtEzQAAikhEIj6kVpv0YjPIFRptEckmUhHoIpYhViaG8EgoBhgIGW+0I9NFaJZ0LRXz8TNJkA5GugAAVOKUCDMZF5TCQAIx6AAMHo8omgsh80Bwkio5Gp+lCZIpVJpIiwhDY6g20BEGFYGQAcoUSogyjJCFgNDRyI0ABStC1dbpSR0yTAAd2gteQWBQyFMP0O-2hGXqHgV8k4ZEFpgAzB4RNZITQIspFnBaNB5NSIYgorPll1CGdoKXOl0MwAyRbb8wlEPxjLDgA00DTjWKdtzI0oAH4j7GANaEDC1zenI-trEu2qK9oCBLwvwZAQcG-RA8CuaBODiSQyDwIFrzTBoAEpWmwyBMK3EgMEkOCoCgcs4FechlHzbMGR9JFoAAcmHRjPFzDBIKlGU5WYVhFxDQIom8aZ5k5RiAKOTJgOvMD2MgnBoMbOCEIpRBkNgtDoGHBpGPVJYoGTDBi2HTD2TATkAHkAGkAEJw0gQziwAVlMsjzMtSjqMCQV6PmJjh2gAAfaAnNY8COPBLjZQYXiFyXQTRl8xQxIkzspJA2SiAiqCYOUxC1JQzSAuCpydL0zUDJTZzIq0wKXLMjojE8mifJEp5GLzAsKDCuTIpQaVovlPj4pXRK2oWS1xMxSSbEBDLwvkxTYPg-L1NQ69OsLMr7Mcpyas2ihXMgMjgFOs7zouy6ruum7bru0jzXcyt+mXGrTEM7ti0wjwqCwBlazUZV5zwXl+U4QUTQCIJ8nK4xLRtUaynEKs2OgZ03Tdd1vTEP0sEoQNg1DcrlnJSlqWQGM4wTTcqGpaB00zB8iHzQsS1aX9tyPA8j2LE8mDPJgL3Qu8sxzC5RlfXdoA-L8fzOf9prS2bpNAhbwQU3KVtUtbNIw+pcK6XD8POQjiLNcjniKRR41g85XiinjxFYLg2GVDjenITlXnjLABHONNsodmKncVV2VQ9+ztV1QyQRQe3+u4mLXAwFE6fOJEfVpygvkMhEo4ZLPY-jgaeOTgAxFMtwzsRDPRLOriAA)

**Related Issues:** 

https://github.com/microsoft/TypeScript/issues/21756

Also, related to my comment here,

https://github.com/microsoft/TypeScript/issues/23689#issuecomment-512114782

-----

**Comment by AnyhowStep** [comment]
If you remove the `N &` part from the parameters and replace `unknown` with `N`, it works.

```ts
/**
 * We should never be able to create a value of this type legitimately.
 * 
 * `ErrorMessageT` is our error message
 */
interface CompileError<ErrorMessageT extends any[]> {
  /**
   * There should never be a value of this type
   */
  readonly __compileError : never;
}
type ErrorA = CompileError<["I am error A"]>;
type ErrorB = CompileError<["I am error B"]>;

declare const errorA : ErrorA;
/**
 * Different compile errors are assignable to each other.
 */
const errorB : ErrorB = errorA;

/**
 * Pretend this is `v1.0.0` of your library.
 */
declare function foo <N extends number> (
  /**
   * This is how we use `CompileError<>` to prevent `3` from being
   * a parameter
   */
  n : (
    (Extract<3, N> extends never ?
    N :
    CompileError<[3, "is not allowed; received", N]>)
  )
) : void;

/**
 * Argument of type '3' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3]>'.
 */
foo(3);
/**
 * OK!
 */
foo(5);
/**
 * Argument of type '3 | 5' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3 | 5]>'.
 */
foo(5 as 3|5);
/**
 * Argument of type 'number' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", number]>'.
 */
foo(5 as number);

///////////////////////////////////////////////////////////////////

/**
 * The same as `foo<>()` but with a different error message.
 * 
 * Pretend this is `v1.1.0` of your library.
 */
declare function bar <N extends number> (
  n : (
    (Extract<3, N> extends never ?
    N :
    CompileError<[3, "is not allowed; received", N]>)
  )
) : void;

/**
 * Expected: Assignable to each other
 * Actual: Assignable to each other
 */
const fooIsAssignableToBar : typeof bar = foo;
const barIsAssignableToFoo : typeof foo = bar;
```

[Playground](http://www.typescriptlang.org/play/#code/PQKhFgCgAIWh1AptAzgCwPYFcA2ATaAO0QDdEAnaAI2QEMqdkAXDaAY3MVqbuhNpxZkGAGbQmaAJYpxATwAOyRgHNJTSQFtuiHLIB0UWNENwABgFFy5DOQCyiFClrLEAFVPRp0bJQrXKGg5OLibAUJKEPOQitGzIAMIYGvKSjJb+ADzpNvaOzm7QiAAePIR4MrSEsgDaALoAfNAA3obQoBAw0EauaBTI6Nj4RKQU1Lz8gsJiEl5MCoitsGGdnLR4GIS60AD622xJKWlWNtAAXMNk5ADcUAC+UHOK0NnkAILQALzQicmpiC8ZaoAIgAktBaBpCsdKK8gQ0bpBHsgXgAhT7fA5-AHAsEQqH+aAouH1BFQPCINg4WicdgbFBMfE2d7nF6vBHtEzQAAikhEIj6kVpv0YjPIFRptEckmUhHoIpYhViaG8EgoBhgIGW+0I9NFaJZ0LRXz8TNJkA5GugAAVOKUCDMZF5TCQAIx6AAMHo8omgsh80Bwkio5Gp+lCZIpVJpIiwhDY6g20BEGFYGQAcoUSogyjJCFgNDRyI0ABStC1dbpSR0yTAAd2gteQWBQyFMP0O-2hGXqHgV8k4ZEFpgAzB4RNZITQIspFnBaNB5NSIYgorPll1CGdoKXOl1i+YSiH4xlhwAaaBpxrFO25kaUAD8iwvZyf7axXeqZ+gQK8hAwDIEHAMEbPArmgTg4kkMg8CBc80waABKVokMgBCtxIDBJFAqAoHLOBXnIZR82zBkfSRaAAHJhwozxc3-cEpRlOVmFYRcQ0CKJvGmeZOQot8jkyT9zx-OiAJwICQLAiDECgxAYPPYcGgo9UligZMMGLYcEPZMBOQAeQAaQAQnDSB1OLABWbTcN0y0CKIwJBTI+ZKOHaAAB9oAsmjf3oyUUGlWUGBYhclw40ZnMUXj+M7QSvxEog-PE4C5KkikZOg2DoDczyLKUlTNTUlNLIY7L3KsnSOiMeziKc7ingovMCwoHzRIYgKmOC8RWLClcIvqhZLT4zEBJsQF4t8sSJNS8D0tk+SiHzQt8tM8yLNKprC2syBcOAPb9oOw6juOk7TrO86cPNWzK36ZdStMdTu2LBCPCoLAGVrNRlXnPBeX5ThBRNAIgnyArjEtG0+rKcQq1o6BnTdN13W9MQ-SwShA2DUMCuWclKWpZAYzjBNNyoaloHTTMb0W5qi23VpN3OHcK23A8mCPJgTzgq8sxzC5RkfXdn1OV8RtisahO-SbwWSyTZsgzK4MQ5CoDQ84MKws08OeIpFHjOTzleRigvlVguDYZV-16chOVeeMsAEc40z843mO6xULZVa3TO1XV1JBFAjY6k23AwFFyfOJEfTJygvnUhFfYZGOA6DwLmNcDAADEUy3KOxHU9EY6uIA)

**Comment by AnyhowStep** [comment]
The motivation behind introducing intersection types is so that multiple compile-time requirements can be chained together,

```ts
declare function bar <N extends number> (
  n : (
    N &
    (Extract<3, N> extends never ? unknown : CompileError<[3, "is not allowed; received", N]>) &
    (Extract<4, N> extends never ? unknown : CompileError<[4, "is not allowed; received", N]>) &
    (SomeOtherComplicatedCondition<N>)
  )
) : void;
```

Of course, with such a "basic" type like `number`, the conditions aren't very interesting.

-----

A code snippet of something more complicated,

```ts
export type AssertNotInPreviousJoinsImpl<
    QueryT extends IQuery,
    AliasedTableT extends IAliasedTable
> = (
    QueryT["_joins"] extends IJoin[] ?
    (
        Extract<
            AliasedTableT["alias"],
            JoinArrayUtil.TableAliases<QueryT["_joins"]>
        > extends never ?
        unknown :
        CompileError<[
            "Alias",
            Extract<
                AliasedTableT["alias"],
                JoinArrayUtil.TableAliases<QueryT["_joins"]>
            >,
            "already used in previous JOINs",
            JoinArrayUtil.TableAliases<QueryT["_joins"]>
        ]>
    ) :
    unknown
);
export type AssertNotInParentJoinsImpl<
    QueryT extends IQuery,
    AliasedTableT extends IAliasedTable
> = (
    QueryT["_parentJoins"] extends IJoin[] ?
    (
        Extract<
            AliasedTableT["alias"],
            JoinArrayUtil.TableAliases<QueryT["_parentJoins"]>
        > extends never ?
        unknown :
        CompileError<[
            "Alias",
            Extract<
                AliasedTableT["alias"],
                JoinArrayUtil.TableAliases<QueryT["_parentJoins"]>
            >,
            "already used in parent JOINs",
            JoinArrayUtil.TableAliases<QueryT["_parentJoins"]>
        ]>
    ) :
    unknown
);
export type AssertNoUsedRefImpl<
    AliasedTableT extends IAliasedTable
> = (
    Extract<keyof AliasedTableT["usedRef"], string> extends never ?
    unknown :
    CompileError<[
        "Derived table",
        AliasedTableT["alias"],
        "must not reference outer query tables",
        Writable<
            ColumnIdentifierUtil.FromColumnRef<AliasedTableT["usedRef"]>
        >
    ]>
);
export type AssertValidJoinTargetImpl<
    QueryT extends IQuery,
    AliasedTableT extends IAliasedTable
> = (
    & AssertNotInPreviousJoinsImpl<QueryT, AliasedTableT>
    & AssertNotInParentJoinsImpl<QueryT, AliasedTableT>
    & AssertNoUsedRefImpl<AliasedTableT>
);
export type AssertValidJoinTarget<
    QueryT extends IQuery,
    AliasedTableT extends IAliasedTable
> = (
    AliasedTableT &
    AssertValidJoinTargetImpl<QueryT, AliasedTableT>
);
```

If all the conditions are satisfied, `AssertValidJoinTarget<QueryT, AliasedTableT>` resolves to `AliasedTableT`.

If at least one of the conditions fail, it resolves to `AliasedTableT & CompileError<>`

**Comment by AnyhowStep** [comment]
It seems like if I nest conditional types instead of introducing intersections, it works,

```ts
/**
 * We should never be able to create a value of this type legitimately.
 * 
 * `ErrorMessageT` is our error message
 */
interface CompileError<ErrorMessageT extends any[]> {
  /**
   * There should never be a value of this type
   */
  readonly __compileError : never;
}
type ErrorA = CompileError<["I am error A"]>;
type ErrorB = CompileError<["I am error B"]>;

declare const errorA : ErrorA;
/**
 * Different compile errors are assignable to each other.
 */
const errorB : ErrorB = errorA;

type IfExtractExtendsNever<A, B, TrueT, FalseT> = (
  Extract<A, B> extends never ? TrueT : FalseT
);
type Condition0<N, TrueT> = (
  IfExtractExtendsNever<
    3, N,
    TrueT,
    CompileError<[3, "is not allowed; received", N]>
  >
);
type Condition1<N, TrueT> = (
  IfExtractExtendsNever<
    4, N,
    TrueT,
    CompileError<[4, "is not allowed; received", N]>
  >
);
type Condition2<N, TrueT> = (
  IfExtractExtendsNever<
    3, N,
    TrueT,
    CompileError<[3, "is not allowed; received", N, "Same as Condition0 but with different error message"]>
  >
);
type Condition3<N, TrueT> = (
  IfExtractExtendsNever<
    4, N,
    TrueT,
    CompileError<[4, "is not allowed; received", N, "Same as Condition1 but with different error message"]>
  >
);
/**
 * Pretend this is `v1.0.0` of your library.
 */
declare function foo <N extends number> (
  /**
   * This is how we use `CompileError<>` to prevent `3` from being
   * a parameter
   */
  n : (
    Condition0<N, Condition1<N, N>>
  )
) : void;

/**
 * Argument of type '3' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3]>'.
 */
foo(3);
/**
 * OK!
 */
foo(5);
/**
 * Argument of type '3 | 5' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", 3 | 5]>'.
 */
foo(5 as 3|5);
/**
 * Argument of type 'number' is not assignable to parameter of type
 * 'CompileError<[3, "is not allowed; received", number]>'.
 */
foo(5 as number);

///////////////////////////////////////////////////////////////////

/**
 * The same as `foo<>()` but with a different error message.
 * 
 * Pretend this is `v1.1.0` of your library.
 */
declare function bar <N extends number> (
  n : (
    Condition2<N, Condition3<N, N>>
  )
) : void;

/**
 * Expected: Assignable to each other
 * Actual: Not assignable to each other
 */
const fooIsAssignableToBar : typeof bar = foo;
const barIsAssignableToFoo : typeof foo = bar;
```

[Playground](http://www.typescriptlang.org/play/#code/PQKhFgCgAIWh1AptAzgCwPYFcA2ATaAO0QDdEAnaAI2QEMqdkAXDaAY3MVqbuhNpxZkGAGbQmaAJYpxATwAOyRgHNJTSQFtuiHLIB0UWNENwABgFFy5DOQCyiFClrLEAFVPRp0bJQrXKGg5OLibAUJKEPOQitGzIAMIYGvKSjJb+ADzpNvaOzm7QiAAePIR4MrSEsgDaALoAfNAA3obQoBAw0EauaBTI6Nj4RKQU1Lz8gsJiEl5MCoitsGGdnLR4GIS60AD622xJKWlWNtAAXMNk5ADcUAC+UHOK0NnkAILQALzQicmpiC8ZaoAIgAktBaBpCsdKK8gQ0bpBHsgXgAhT7fA5-AHAsEQqH+aAouH1BFQPCINg4WicdgbFBMfE2d7nF6vBHtEzQAAikhEIj6kVpv0YjPIFRptEckmUhHoIpYhViaG8EgoBhgIGW+0I9NFaJZ0LRXz8TNJiPm0BBInMJXIsSYNtK5QAciNyBlXgAaQne1zkISub0AMQEKDcjS+AApWo67WwmB7vSjGsUnTJiJdoAB+aB+gNnaAhnBh1xQACUCKRGLKakkGwADBlnb7-eH0dHOlbY-bHYgyihXZcMotoABmb3Nkd5tyekc-Q7-aGA8fQIFeQgYBkCHAYADuiDwV2gnDikjIeCBE4arXq5crFsSNfUGwAjE2WwGI9AO10u7aeyUfYum6w6dF0AAsE6zmBuatoGc6YkcmTVJBq7rpu4I4Du+6HseFKIGeB6XtAzrXp0t6QBWDwPhseC1hsABM76wZ+7atH+TBxg6gH9oOFCgV0XQrpOMHTvBMHzliS7VCua7phh257geR4ngR57Ec2q4AMoQnQMiPnRz6EPW1BYAyu5qMqdF8gKDImgEQT5MSN53tRTwGfRhCjsxYlfj+lrWv+8a9rxIEjqhImCSxM4IcKi7IahclEApWFKbhqmEReUHabp4L6bRnkvqZ5mWdA1n8pwgr2dAgR5C4znka5kAchq0AAAqcE64hSDIXimCQL56PWQ0eKI0CyD40A4JIVB2uQ+ihGSFJUjSIhYIQ8Z1oQ0AiBgrBNoUPHlEQWAaDQ5CNP5LWCXAPReF4mC7tA+7QFgYbQKYklITYGT1B4CryJwZCCqYo4eCI1iQjQETKIscC0NA8jUrpURw8sXTbec-ldB5RmNppuNbW+mnOvUFFdGW5YFiQGCSIeUBQNdcCvOQyinX2DJjVWADko7c548lblKMpyswrBI3agRRN40zzJy3NffFP0yd6SUbluqU4Sp+GZcRo4NNz6pLFAu0YJGo5Uc1YCcgA8gA0gAhItkCm5GACsltM9ALNs4EgpcxavPQAAPtAbv8+hQsoNKsoMGLiPI1LowB4o8uK9ismR5h2HKXhp7qd6o4h2HBtG5qJt7e7eVjsHHvstbrU++z-uy083OEKd50R4LeXRyLcfiOLieINLKcLK1CuIUr7oq2hPeKVredqUR3od2dFCl87rtu9Xa-nZbjPAEfx8n6fZ-nxfl9X9fDNWx03S9KguWSh9pu-ZGZYeFQZnPaVCPlbZUUNVHIuDLsYVqHUR5AW6vdGQ-VBqDXrKNMQE0sCUGmrNakC0NTLHJJSakyA1obSMtQak0ADqpiAumTuFBLqtExt+BCT4tpMQJgVIy3kSZk1aJTSi1Nab00gIzBuRgbSKHjAec4rxhax3lKwLgbBlSbl6OQTkrx4xYAEOcZ0CkZGi0HoqRRKoVHO21LqU2IIUDSL7rItwGAURkPOEiMaVAyFfFNgiMxDJXHkEsdYmOotXAYCDHtAszixCm3RD4q4QA)

**Comment by AnyhowStep** [comment]
This assignability situation is actually making life kind of hard ._.

```ts
/**
 * We should never be able to create a value of this type legitimately.
 * 
 * `ErrorMessageT` is our error message
 */
interface CompileError<ErrorMessageT extends any[]> {
  /**
   * There should never be a value of this type
   */
  readonly __compileError : never;
}
type ErrorA = CompileError<["I am error A"]>;
type ErrorB = CompileError<["I am error B"]>;

declare const errorA : ErrorA;
/**
 * Different compile errors are assignable to each other.
 */
const errorB : ErrorB = errorA;

/**
 * Expected:
 * OK!
 * 
 * Actual:
 * Type
 * '<N extends number>(n: N & (Extract<3, N> extends never ? unknown : CompileError<[3, "is not allowed; received", N]>)) => void'
 * is not assignable to type
 * '<N extends number>(n: N & (Extract<3, N> extends never ? unknown : CompileError<[3, "is not allowed; received", N]>)) => void'.
 * Two different types with this name exist, but they are unrelated.
 */
const blah : (
  <N extends number> (
    n : (
      N &
      (Extract<3, N> extends never ?
      unknown :
      CompileError<[3, "is not allowed; received", N]>)
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (Extract<3, N> extends never ?
      unknown :
      CompileError<[3, "is not allowed; received", N]>)
    )
  ) : void => {

  }
)
```

[Playground](http://www.typescriptlang.org/play/#code/PQKhFgCgAIWh1AptAzgCwPYFcA2ATaAO0QDdEAnaAI2QEMqdkAXDaAY3MVqbuhNpxZkGAGbQmaAJYpxATwAOyRgHNJTSQFtuiHLIB0UWNENwABgFFy5DOQCyiFClrLEAFVPRp0bJQrXKGg5OLibAUJKEPOQitGzIAMIYGvKSjJb+ADzpNvaOzm7QiAAePIR4MrSEsgDaALoAfNAA3obQoBAw0EauaBTI6Nj4RKQU1Lz8gsJiEl5MCoitsGGdnLR4GIS60AD622xJKWlWNtAAXMNk5ADcUAC+UHOK0NnkAILQALzQicmpiC8ZaoAIgAktBaBpCsdKK8gQ0bpBHsgXgAhT7fA5-AHAsEQqH+aAouH1BFQPCINg4WicdgbFBMfE2d7nF6vBHtEzQAAikhEIj6kVpv0YjPIFRptEckmUhHoIpYhViaG8EgoBhgIGW+0I9NFaJZ0LRXz8TNJkA5GueRUUbB4eFOnIA8gBpACEnM5r1tWAEDstrnmnIA5BkAHKFEqIMoyQhYDQ0cj1AAUhHO4YAZNAk+YSuRYkwMgBmAA00FDjWKpXKF1GAH5oFhCABrQgYADuhDOGOF-2hgJL0CBXlbDIEOHbiDwV2gnDikjIeCBpdDDQAlKvPo0SBhJHgg5zhxhR1KZXLmKwkcGwxGqzG4wnk6my9BM9nc-mi8uK5HozXKPXGxbdtO3OH5Dl7TJqgHIcYyPcEcHHNtJ2nWdEHnSclzLNcNw+Lcdz3dVujbVg8F5flOEFJEZDbNRlRmGMIWQYppCYUsqCwBlVVkcEaUbTgqTtQjNSgbVdQYWhlXOJNWmvSso2rWN4woRppM6LoQKzRYumfdMtK6N8mDzW1PzLb9bz-aBaz0htm1bDszmssCsT7KDSxgog4LHCcpxnCk0IXTCV3qVctJCzocLw3coBwzTOlkn8FPvZTYu0ogu1U1KugzayDKMgsB3LG95JjEZ-2swC7JAxzMSOSDoMPUcEO8lC-PQxdlzXULWg3c5t13TdmigVp7kgVcgA)

-----

In the above, you see something like `A` is not assignable to `A`. And that just confuses me.

My actual use-case has to do with having function properties (not methods) on a generic class.

And the function uses the `this` keyword.

**Comment by AnyhowStep** [comment]
Smaller repro,

```ts

/**
 * Expected:
 * OK!
 * 
 * Actual:
 * Type
 * '<N extends number>(n: N & (3 extends N ? any : any)) => void'
 * is not assignable to type
 * '<N extends number>(n: N & (3 extends N ? any : any)) => void'.
 * Two different types with this name exist, but they are unrelated.
 */
const blah : (
  <N extends number> (
    n : (
      N &
      (3 extends N ? any : any)
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (3 extends N ? any : any)
    )
  ) : void => {

  }
)
```

[Playground](http://www.typescriptlang.org/play/#code/LAKA9AVBoAQTBRAHgBwKYGMAuaAmAuWeAeQGkBCImKgQWwFcBDAG0JDhgBUBPdKgcgA8AORhokOAHa4AzjEn0AtgCM0AJwB8ACkn4YogGQwtAZjES00uaID8MRpO4w9D7gEo3MALwaYANwB7AEtcfioguUkArHsZGSCAc0lGZWY0GCwAjN40ARFzKVl5JVVNHT1DYzNxQusYO1dne0cPb19AkP4AOipOAHcs3CCAM2H1SxisHLk+oKwACwz5iPlGRXTxCKwAGhhlekn5tCdGNXT6STPmRhxcHvYIMFAMAMkZGNTGRb0tWBh8mqWIoKFTqXy-dgwKGSJoQqHw-QwAx-BGmApAuoNRxNVxuFEwPGQzw+fzBXCgYnGP4AixWYqgzRUyHQ2H4qGGNlVdF02zNJwuFr4wlQzx6Dq4NowADeoD+AF8KUA)

**Comment by AnyhowStep** [comment]
```ts
type BLAH<N> = 3 extends N ? any : any;
/**
 * Expected:
 * OK!
 * 
 * Actual:
 * Types of parameters 'n' and 'n' are incompatible.
 * Type 'N & BLAH<N>' is not assignable to type 'N & BLAH<N> & BLAH<N & BLAH<N>>'.
 * Type 'N & BLAH<N>' is not assignable to type 'BLAH<N & BLAH<N>>'.
 */
const blah : (
  <N extends number> (
    n : (
      N &
      BLAH<N>
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      BLAH<N>
    )
  ) : void => {

  }
)
```
[Playground](http://www.typescriptlang.org/play/#code/C4TwDgpgBAQgMgQQBIB4ByA+KBeKBmKCAD2AgDsATAZyjSgH4oBDMkKALmdYG4BYAKAD0AKmECowqAFEikAMakK7cZIDyAaQCEKqDoQKArkwA2y-hKgAVcBBoB7AGZQwTAE5MAthFKuaAcjI-LgooAKC3aABLMjk7DxdgSIAjYwgAOh1rSFC6ADJYRFRMIMiaMjtgZioqSIBzMiYU6GA7KFBsvzyC5HQsfPgeroGijAw-DPNJLOhOqH7C3pKyiqqa+sbUttb2meH0Oe6RsYmJQQFYsipKlKYACw4oAApxKH3iUkoygw8kiFcsZ7mKDAsgPQHAiG0OYvSF7TAwqAASheiJwWAAbnZIhQBKjcODXnR3uRqFAyN9fv8ngjQZwCZC8gjgXCMAjkUDUZxMdi0VAAN4CF4AX1xQA)

**Comment by AnyhowStep** [comment]
So, if I put the generic type param on the `RHS` of the `extends`, it is **not** assignable,
```ts
//Error, cannot assign
const blah : (
  <N extends number> (
    n : (
      N &
      (any extends N ? any : any)
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (any extends N ? any : any)
    )
  ) : void => {

  }
)
```

-----

If I put it on the `LHS` of the `extends`, it **is** assignable,

```ts
//OK! Can assign
const blah : (
  <N extends number> (
    n : (
      N &
      (N extends 3 ? unknown : ["blah"])
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (N extends 3 ? unknown : ["blah"])
    )
  ) : void => {

  }
)
```

-----

And I have zero intuition for why this is the case.

**Comment by RyanCavanaugh** [comment]
@AnyhowStep TL;DR ?

**Comment by AnyhowStep** [comment]
Umm... Kind of hard to make a TL;DR.

-----

### TL;DR

I have a generic function.
It has a parameter.
That parameter uses a conditional type and intersection type.

Because of the conditional and intersection type, a different function with the exact same signature is treated as a different type... Even though they are the same type. A type assertion does not work, either.

I don't know if it is a bug, design limitation, or by design.

And I don't have an intuition for why it thinks it is assignable in some cases, but not assginable in other cases.

-----

```ts
//Error, cannot assign
const blah : (
  <N extends number> (
    n : (
      N &
      (any extends N ? any : any)
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (any extends N ? any : any)
    )
  ) : void => {

  }
)
```
[Playground](http://www.typescriptlang.org/play/#code/PTCiCdwe3AaACAxgQwHaqgF3sgzrgSwHNUBYAKESlV2wCMAbZAC3gC54AKC+eAHgBy8AKYAPTMNQATXPFQBXALZ1h4AHxcevOe03lt2oQDItBzmgCeI8ZJnwhAfhyorHSwEpT8T-u-wAvBoAblAEUhTuAXq8gtYS0rIKyqoa3L68qLppBrzGXrzmLnG2so7OruU+BlV+HCFhARoA3hRaAL4RQA)

Notice that the type annotation and the type of the function have the same "type". But the compiler thinks they're different types.

-----

But I have found that this is not always the case. In certain cases, the compiler knows they're the same type.

Like when I put `N` on the `LHS` of the `extends`.

```ts
//OK! Can assign
const blah : (
  <N extends number> (
    n : (
      N &
      (N extends 3 ? unknown : ["blah"])
    )
  ) => void
) = (
  <N extends number> (
    n : (
      N &
      (N extends 3 ? unknown : ["blah"])
    )
  ) : void => {

  }
)
```
[Playground](http://www.typescriptlang.org/play/#code/PTDyGkEIAIGEEMB214GdUEsDmiCwAoAYwHtFUAXaAIwBt4ALaALmgAoDpoAeAOWgFMAHuX6IAJqmiIArgFsq-AE4A+Nh05Tma-Bo18AZOt2s+QkeMkBmaAH5o0xAGtExAO7IWAbQBEtBt4BdAEojaBCdMOgAXlUAN2IMMQIg6O1OXgFhUQkpOQUVNI0PQt0DUM4TTPMc6zsHZzdinz96QPDddsiWeMTo1QBvAnUAX2SgA)
